ambry-clustermap/src/main/java/com/github/ambry/clustermap/CompositeDataNodeConfigSource.java

```diff
@@ -0,0 +1,190 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ */
+
+package com.github.ambry.clustermap;
+
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link DataNodeConfigSource} that combines two sources. The "primary" acts as the source of
+ * truth for users of this class, but results obtained are compared against the "secondary" source. This class can be
+ * useful for safe migrations between different backing stores.
+ */
+public class CompositeDataNodeConfigSource implements DataNodeConfigSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CompositeDataNodeConfigSource.class);
+  static final long CHECK_PERIOD_MS = TimeUnit.MINUTES.toMillis(5);
+  static final int MAX_PERIODS_WITH_INCONSISTENCY = 3;
+  private final DataNodeConfigSource primarySource;
+  private final DataNodeConfigSource secondarySource;
+  private final DataNodeConfigSourceMetrics metrics;
+  private final Time time;
+  private final AtomicBoolean recordingListenersRegistered = new AtomicBoolean(false);
+
+  /**
+   * @param primarySource the source of truth for users of this class.
+   * @param secondarySource all operations are also attempted against this source, but if the results of the operation
+   *                        differ from the primary, inconsistencies are just recorded with logs and metrics.
+   * @param time a {@link Time} instance used for consistency verification.
+   * @param metrics a {@link DataNodeConfigSourceMetrics} instance to use.
+   */
+  public CompositeDataNodeConfigSource(DataNodeConfigSource primarySource, DataNodeConfigSource secondarySource,
+      Time time, DataNodeConfigSourceMetrics metrics) {
+    this.primarySource = primarySource;
+    this.secondarySource = secondarySource;
+    this.time = time;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public void addDataNodeConfigChangeListener(DataNodeConfigChangeListener listener) throws Exception {
+    if (recordingListenersRegistered.compareAndSet(false, true)) {
+      ConfigChangeRecorder primaryListener = new ConfigChangeRecorder(listener);
+      ConfigChangeRecorder secondaryListener = new ConfigChangeRecorder(null);
+      primarySource.addDataNodeConfigChangeListener(primaryListener);
+      try {
+        secondarySource.addDataNodeConfigChangeListener(secondaryListener);
+        Utils.newThread("DataNodeConfigConsistencyChecker",
```

Comment:
Curious .. Why do we need to create a periodic thread to check consistency? Why can't we check on each call to `ConfigChangeRecorder.onDataNodeConfigChange` which is where we seem to be setting `ConfigChangeRecorder.configsSeen`.

Reply:
I too would prefer to lazily check when `onDataNodeConfigChange` is called and actually implemented it at first before switching to the background thread approach. I saw an issue in the following scenario:
1. Listener 1 gets a notification. This indicates that there is a transient inconsistency but we don't want to alert yet since the update for listener 2 could come before or after.
2. Listener 2 just never gets a notification. Without a background thread to check periodically, nothing will alert until another listener notification comes, potentially far in the future.

If we assume that notifications come frequently enough (and a situation where both listeners stop receiving notifications is rare), we could probably go back to this approach.